<a href="https://colab.research.google.com/github/HQhanqiZHQ/DL_labs/blob/main/ProblemSet1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem Set 1
### Due Wednesday, April 3rd by 11:59pm EST
**Your Harvard ID**: 81631239

Note: when completed, please download as an IPython Notebook file and upload to Canvas.

### Question 1
You will be performing one iteration of the forward pass and backpropagation calculations for a small network using Python. Here we will focus on the calculations for one training example, though in reality your data sets will be much larger and require matrix computation. You will also calculate the associated loss.

Let $X_1 = 2$ and $X_2 = -1$ be the feature inputs and initialize the weights to be as shown in the figure below. This is a neural network with a single hidden layer consisting of three nodes. The blue numbers within each node represent the values for the bias terms and the black numbers along the edges represent the weights. The hidden layer outputs a single node, from which your task is binary classification. The label for this particular training example outcome is $y = 1$.



![network](https://drive.google.com/uc?id=1lfmGA56cIu81xD0y1SPKB7VuddS11G5o)


Implement a single forward pass of the network. Assume the hidden layer uses a linear activation function (which is equivalent to assuming no activation function). You do not need to implement the network in keras and should instead use numpy operations (either scalar or matrix). Please use the variable names and print statements provided in the code chunks to display results for the TAs.

In [2]:
# Your code here
import numpy as np
# Initialize the inputs and weights
X = np.array([2, -1])
W_hidden = np.array([[1, 1.1], [0.2, 0], [-0.6, -0.3]])
b_hidden = np.array([-1.8, -0.4, 0.96])
W_output = np.array([0.5, 0.1, 1.3])
b_output = 2

# Perform the forward pass
hidden = np.dot(W_hidden, X) + b_hidden
output = np.dot(W_output, hidden) + b_output
y_hat = 1 / (1 + np.exp(-output))  # Sigmoid for binary classification
prediction = 1 if y_hat >= 0.5 else 0

print('The values for the hidden layer are:', hidden)
print('The value for the output layer is:', output)
print('The predicted probability is:', y_hat)
print('The prediction is:', prediction)

The values for the hidden layer are: [-0.9   0.    0.06]
The value for the output layer is: 1.6280000000000001
The predicted probability is: 0.8358954745887476
The prediction is: 1


Calculate the loss for the training example making sure to select the appropriate loss function.

In [3]:
# Your code here
# Given true label
y_true = 1

# Compute the loss using binary cross-entropy
loss_i = - (y_true * np.log(y_hat) + (1 - y_true) * np.log(1 - y_hat))

print('The loss is:',loss_i)

The loss is: 0.17925170411062194


Implement a single backward pass of the network. Again use numpy and report the values using the print statements provided. Please interpret these values. In other words, what are the values you just calculated used for?

In [4]:
# Your code here
# Forward pass was already computed; we are reusing the hidden, output, and y_hat variables.
# Let's assume these variables are populated with the results from the forward pass we calculated earlier.

# Backward pass: Compute the gradient of the loss w.r.t. the network's parameters
d_loss_output = y_hat - y_true  # Gradient of loss w.r.t. output of network (pre-activation)
d_loss_hidden = W_output * d_loss_output  # Gradient of loss w.r.t. hidden layer output

# Gradient of loss w.r.t. hidden layer weights and biases
d_loss_W_hidden = np.outer(d_loss_hidden, X)  # d_loss_hidden reshaped for outer product with X
d_loss_b_hidden = d_loss_hidden  # Gradient of loss w.r.t. hidden layer bias is the same as d_loss_hidden

# Gradient of loss w.r.t. output layer weights and bias
d_loss_W_output = hidden * d_loss_output  # Element-wise multiplication
d_loss_b_output = d_loss_output  # Gradient of loss w.r.t. output layer bias

dl_dw_h = d_loss_W_output
dl_db_h = d_loss_b_output

dl_dw_1 = d_loss_W_hidden[0]
dl_dw_2 = d_loss_W_hidden[1]
dl_dw_3 = d_loss_W_hidden[2]

print('The gradients of the loss wrt to the hidden weights are:', dl_dw_h)
print('The gradient of the loss wrt to the hidden bias is:', dl_db_h)
print('The gradients of the loss wrt to the input weights going to hidden node 1 are:', dl_dw_1)
print('The gradients of the loss wrt to the input weights going to hidden node 2 are:', dl_dw_2)
print('The gradients of the loss wrt to the input weights going to hidden node 3 are:', dl_dw_3)

The gradients of the loss wrt to the hidden weights are: [ 0.14769407 -0.         -0.00984627]
The gradient of the loss wrt to the hidden bias is: -0.16410452541125242
The gradients of the loss wrt to the input weights going to hidden node 1 are: [-0.16410453  0.08205226]
The gradients of the loss wrt to the input weights going to hidden node 2 are: [-0.03282091  0.01641045]
The gradients of the loss wrt to the input weights going to hidden node 3 are: [-0.42667177  0.21333588]


The gradients of the loss wrt to the hidden weights are: [ 0.14769407 -0.         -0.00984627] \
The gradient of the loss wrt to the hidden bias is: -0.16410452541125242 \

The gradients of the loss wrt to the input weights going to hidden node 1 are: [-0.16410453  0.08205226] \
The gradients of the loss wrt to the input weights going to hidden node 2 are: [-0.03282091  0.01641045] \
The gradients of the loss wrt to the input weights going to hidden node 3 are: [-0.42667177  0.21333588] \
A negative gradient implies that in order to minimize the loss, we should increase the corresponding weight in the next iteration.
A positive gradient implies that we should decrease the weight to minimize the loss.\
For hidden weights (associated with the gradient 0.14769407 -0.         -0.00984627), implies that in order to minimize the loss, we should increase, decrease, and decrease the corresponding weights between hidden layers and output layers. \

The gradient of the loss with respect to the hidden bias is negative(-0.16410452541125242), implying in order to minimize the loss, we should decrease the bias. \

For hidden node 1:

The weight from the first input feature (associated with the gradient -0.16410453) should be increased.
The weight from the second input feature (associated with the gradient 0.08205226) should be decreased. \
For hidden node 2:

The weight from the first input feature (associated with the gradient -0.03282091) should be increased.
The weight from the second input feature (associated with the gradient 0.01641045) should be decreased. \
For hidden node 3:

The weight from the first input feature (associated with the gradient -0.42667177) should be increased.
The weight from the second input feature (associated with the gradient 0.21333588) should be decreased.

The magnitude of these gradients gives an indication of how "sensitive" or "influential" a particular weight is with respect to the loss:

Large negative gradients suggest that increasing the weight could potentially lead to a significant decrease in the loss.
Small negative gradients indicate a lesser degree of sensitivity, implying that changes to the weight might result in smaller decreases in the loss.
Large positive gradients suggest that decreasing the weight could lead to a significant decrease in the loss.
Small positive gradients indicate that the weight has less influence on reducing the loss when it is decreased.

Gradient with respect to the hidden weights indicate how much to adjust the weights for the output node. \
Gradient with respect to the hidden bias indicate how much to adjust the bias for the output node. \
For the first hidden node bias (with a gradient of -0.08205226), because the gradient is negative, it suggests that increasing the bias will help to decrease the loss. This can be thought of as adjusting the threshold at which the neuron activates, allowing it to fire more easily. \
For the second hidden node bias (with a gradient of -0.01641045), this value is also negative, but it's smaller in magnitude than the first node's gradient. This suggests that while the bias should still be increased to minimize the loss, the impact of this particular bias on the loss is less compared to the first node. \
For the third hidden node bias (with a gradient of -0.21333588), this gradient is the most negative among the three. This indicates a relatively larger potential reduction in loss by increasing the bias for this node. It may suggest that this node has a more substantial role or is more sensitive in the prediction errors being produced by the current model.




### Question 2
In class we were considering classification problems where the goal was to predict a single discrete label of an input data point. Another common type of machine learning problem is "regression", which consists of predicting a continuous value instead of a discrete label. For instance, predicting the temperature tomorrow, given meteorological data, or predicting the time that a software project will take to complete, given its specifications.

You will be attempting to predict the median price of homes in a given Boston suburb in the mid-1970s, given a few data points about the suburb at the time, such as the crime rate, the local property tax rate, etc.

The dataset you will be using has another interesting difference from our previous examples: it has very few data points, only 506 in total, split between 404 training samples and 102 test samples, and each "feature" in the input data (e.g. the crime rate is a feature) has a different scale. For instance, some values are proportions, which take a value between 0 and 1, others take values between 1 and 12, others between 0 and 100.

The data consists 13 features. The 13 features in the input data are as follows:

1. Per capita crime rate.
2. Proportion of residential land zoned for lots over 25,000 square feet.
3. Proportion of non-retail business acres per town.
4. Charles River dummy variable (= 1 if tract bounds river; 0 otherwise).
5. Nitric oxides concentration (parts per 10 million).
6. Average number of rooms per dwelling.
7. Proportion of owner-occupied units built prior to 1940.
8. Weighted distances to five Boston employment centres.
9. Index of accessibility to radial highways.
10. Full-value property-tax rate per $10,000.
11. Pupil-teacher ratio by town.
12. 1000(Bk - 0.63)^2 where Bk is the proportion of Black people by town.
13. % lower SES status of the population.

The targets (outcomes, y) are the median values of owner-occupied homes, in thousands of dollars. The prices are typically between 10,000 and 50,000 dollars. If that sounds cheap, remember this was the mid-1970s, and these prices are not inflation-adjusted.

In [ ]:
# Import needed packages
import tensorflow as tf
import numpy as np
from tensorflow import keras
import pandas as pd
from tensorflow.keras.datasets import reuters
from keras.utils import to_categorical
from tensorflow.keras import layers
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
# Load the data
from keras.datasets import boston_housing

(train_data, train_targets), (test_data, test_targets) =  boston_housing.load_data()

57026/57026 [==============================] - 0s 0us/step


Print the dimensions of the training set, i.e. its shape

In [ ]:
# Your code here
print(train_data.shape)

(404, 13)


Print the dimensions of the test set, i.e. its shape

In [ ]:
# Your code here
print(test_data.shape)

(102, 13)


It would be problematic to feed into a neural network values that all take wildly different ranges. The network might be able to automatically adapt to such heterogeneous data, but it would definitely make learning more difficult. A widespread best practice to deal with such data is to do feature-wise normalization: for each feature in the input data (a column in the input data matrix), we will subtract the mean of the feature and divide by the standard deviation, so that the feature will be centered around 0 and will have a unit standard deviation.

Normalize the data. Be sure to normalize the test set with the training set mean and standard deviation.

In [ ]:
# Your code here
mean = train_data.mean(axis=0)
std = train_data.std(axis=0)
train_data -= mean
train_data /= std
test_data -= mean
test_data /= std

Fit a fully connected neural network with 2 hidden layers and an output layer. Include 64 hidden units in each hidden layer and an appropriate number of units in the output layer. You are free to choose the activation functions. Use the `rmsprop` optimization function, and choose an appropriate loss function and model performance measure. Referring to the table shown in lectures 2 and 3 may help with these choices. Run the network for 50 epochs and use a batch size of 10.

In [ ]:
# Your code here
def build_model():
    model = keras.Sequential([
        layers.Dense(64, activation='relu', input_shape=(train_data.shape[1],)),
        layers.Dense(64, activation='relu'),
        layers.Dense(1)  # Single output node for regression
    ])
    model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])
    return model

model = build_model()
model.fit(train_data, train_targets, epochs=50, batch_size=10, verbose=1)

Epoch 1/50
41/41 [==============================] - 1s 2ms/step - loss: 427.3323 - mae: 18.5177
Epoch 2/50
41/41 [==============================] - 0s 2ms/step - loss: 161.4739 - mae: 10.2613
Epoch 3/50
41/41 [==============================] - 0s 2ms/step - loss: 58.8808 - mae: 5.7144
Epoch 4/50
41/41 [==============================] - 0s 2ms/step - loss: 33.8611 - mae: 4.1696
Epoch 5/50
41/41 [==============================] - 0s 2ms/step - loss: 25.9387 - mae: 3.5879
Epoch 6/50
41/41 [==============================] - 0s 2ms/step - loss: 21.9441 - mae: 3.2773
Epoch 7/50
41/41 [==============================] - 0s 2ms/step - loss: 19.1014 - mae: 3.0551
Epoch 8/50
41/41 [==============================] - 0s 2ms/step - loss: 16.8161 - mae: 2.8940
Epoch 9/50
41/41 [==============================] - 0s 2ms/step - loss: 15.0817 - mae: 2.7531
Epoch 10/50
41/41 [==============================] - 0s 2ms/step - loss: 14.3921 - mae: 2.6530
Epoch 11/50
41/41 [==============================] - 0s

Report the test set accuracy (as in, a measure of how good your predictions are) and compare it to the training set accuracy. **Interpret what this means in words, in terms of what you are trying to do with your network**.

In [ ]:
# Your code here
test_mse_score, test_mae_score = model.evaluate(test_data, test_targets)
print(f"Test MSE with 64 hidden units: {test_mse_score}\nTest MAE with 64 hidden units: {test_mae_score}")

4/4 [==============================] - 0s 4ms/step - loss: 18.5742 - mae: 2.6495
Test MSE with 64 hidden units: 18.574249267578125
Test MAE with 64 hidden units: 2.649498224258423


Answer: Test MSE with 64 hidden units is 18.574249267578125.
Test MAE with 64 hidden units is 2.649498224258423. An MSE of 18.5742 means that the average squared difference between the predicted and actual home prices (in thousands of dollars, squared) is around 18.5742. With an MAE of 2.6495, this means that on average, the model's predictions are off by about $2,649.50.

Now fit the same network as above but with 16 hidden nodes in each hidden layer. **What is the test set accuracy and how does it compare to the first network you created? Which model do you think is better?**

In [ ]:
# Your code here
def build_model_with_16():
    model = keras.Sequential([
        layers.Dense(16, activation='relu', input_shape=(train_data.shape[1],)),
        layers.Dense(16, activation='relu'),
        layers.Dense(1)
    ])
    model.compile(optimizer='rmsprop', loss='mse', metrics=['mae'])
    return model

model_with_16 = build_model_with_16()
model_with_16.fit(train_data, train_targets, epochs=50, batch_size=10, verbose=1)
test_mse_score_16, test_mae_score_16 = model_with_16.evaluate(test_data, test_targets)
print(f"Test MSE with 16 hidden units: {test_mse_score_16}\nTest MAE with 16 hidden units: {test_mae_score_16}")

Epoch 1/50
41/41 [==============================] - 1s 2ms/step - loss: 554.5681 - mae: 21.8409
Epoch 2/50
41/41 [==============================] - 0s 2ms/step - loss: 495.8397 - mae: 20.5566
Epoch 3/50
41/41 [==============================] - 0s 2ms/step - loss: 423.4792 - mae: 18.8721
Epoch 4/50
41/41 [==============================] - 0s 2ms/step - loss: 338.9770 - mae: 16.6196
Epoch 5/50
41/41 [==============================] - 0s 2ms/step - loss: 246.7624 - mae: 13.7058
Epoch 6/50
41/41 [==============================] - 0s 2ms/step - loss: 163.6617 - mae: 10.3775
Epoch 7/50
41/41 [==============================] - 0s 2ms/step - loss: 103.1222 - mae: 7.7468
Epoch 8/50
41/41 [==============================] - 0s 2ms/step - loss: 70.3271 - mae: 6.1411
Epoch 9/50
41/41 [==============================] - 0s 2ms/step - loss: 53.8182 - mae: 5.1874
Epoch 10/50
41/41 [==============================] - 0s 2ms/step - loss: 43.5869 - mae: 4.5190
Epoch 11/50
41/41 [===========================

Answer: Test MSE with 16 hidden units is 25.357641220092773, larger than 18.574249267578125 which is the Test MSE with 64 hidden units. Test MAE with 16 hidden units is 3.220874786376953, larger than 2.649498224258423 which is the Test MAE with 64 hidden units. These suggest that the model with 64 hidden units is a better fit for this particular dataset since lower values of MSE and MAE indicate a model that is making predictions closer to the true values, which is preferred. However, this conclusion is drawn mostly based on the test results, and in real situations, sometimes a model with fewer parameters (like the one with 16 hidden units per layer) may generalize better and be less prone to overfitting, especially if the dataset is small.